### PR TITLE
Refactor Compiler

### PR DESCRIPTION
**Pull request by giginet**
For next feature development, first, refactor `Compiler`.

I added `FrameworkProducer` to produce XCFramework.
This producer uses `Compiler` and `CacheSystem`.

### CODE REVIEW

File: Sources/ScipioKit/Producer/Compiler.swift [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+            return DebugSymbol(dSYMPath: dsymPath,
+                               target: target,
+                               sdk: sdk,
+                               buildConfiguration: buildConfiguration)
+        }
+        // You can use AsyncStream
+        var symbolMapPaths: [URL] = []
+        for dSYMs in debugSymbols {
+            let maps = try await self.extractor.dump(dwarfPath: dSYMs.dwarfPath)
+            let paths = maps.values.map { uuid in
```
Comment (freddi-kit) [inline Sources/ScipioKit/Producer/Compiler.swift:87]: I prefer to use more specific name than map

```suggestion
            let dumpedDSYMsMaps = try await self.extractor.dump(dwarfPath: dSYMs.dwarfPath)
            let paths = dumpedDSYMsMaps.values.map { uuid in
```

File: Sources/ScipioKit/Producer/FrameworkProducer.swift [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+                                      buildOptions: buildOptions,
+                                      outputDirectory: outputDir,
+                                      storage: cacheStorage)
+        let compiler = Compiler(rootPackage: rootPackage, buildOptions: buildOptions)
+        try await compiler.clean()
+
+        let products = productsToBuild(for: mode)
+        for product in products {
+            if product.isBinaryTarget {
+                logger.error("🚧 \(product.target) is binaryTarget. Currently, binaryTarget is not supported. Skip it.")
```
Comment (freddi-kit) [inline Sources/ScipioKit/Producer/FrameworkProducer.swift:48]: I think warning is better

File: Sources/ScipioKit/Runner.swift [hunk 3]
```diff
@@ -62,17 +62,6 @@ public struct Runner {
             case disabled
             case project
             case storage(any CacheStorage)
-
```
Comment (freddi-kit) [inline Sources/ScipioKit/Runner.swift:65]: 👍

File: Sources/ScipioKit/Producer/FrameworkProducer.swift [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+                return
+            } else {
+                logger.warning("⚠️ Existing \(frameworkName) is outdated.", metadata: .color(.yellow))
+                logger.info("💥 Delete \(frameworkName)", metadata: .color(.red))
+                try fileSystem.removeFileTree(at: outputPath)
+                let restored = await cacheSystem.restoreCacheIfPossible(product: product)
+                needToBuild = !restored
+                if restored {
+                    logger.info("✅ Restore \(frameworkName) from cache storage", metadata: .color(.green))
+                }
```
Comment (freddi-kit) [inline Sources/ScipioKit/Producer/FrameworkProducer.swift:96]: Why can we restore cache even the cache is invalid?

File: Sources/ScipioKit/Producer/FrameworkProducer.swift (hunk 2 above)
Comment (giginet) [reply Sources/ScipioKit/Producer/FrameworkProducer.swift:48]:
```suggestion
                logger.warning("🚧 \(product.target) is binaryTarget. Currently, binaryTarget is not supported. Skip it.")
```

File: Sources/ScipioKit/Producer/FrameworkProducer.swift (hunk 4 above)
Comment (giginet) [reply Sources/ScipioKit/Producer/FrameworkProducer.swift:96]: The invalid caches exist on an output dir ( e.g. `$PROJECT_DIR/XCFrameworks`)

On the other hand, Scipio tries to restore from other cache dirs (e.g. `~/Library/Caches`).

If XCFrameworks in output dirs are invalid, then they will be restored from cache dir.
So I think this implementation is correct.

**Review by freddi-kit — APPROVED** [review]

**Pull request merged**